SquirrelPanel.m

```diff
@@ -9,108 +9,607 @@
     }
-    [self.highlightedStripColor setFill];
-    if (self.hilitedCornerRadius > 0) {
-      [[NSBezierPath bezierPathWithRoundedRect:stripRect
-                                       xRadius:self.hilitedCornerRadius
-                                       yRadius:self.hilitedCornerRadius] fill];
+    NSRect outerBox = preeditRect;
+    outerBox.size.height -= _hilitedCornerRadius;
+    outerBox.size.width -= _hilitedCornerRadius;
+    outerBox.origin.x += _hilitedCornerRadius / 2;
+    outerBox.origin.y += _hilitedCornerRadius / 2;
+    
+    NSMutableArray<NSValue *> *highlightedPreeditPoints;
+    NSMutableArray<NSValue *> *highlightedPreeditPoints2;
+    // Handles the special case where containing boxes are separated
+    if (nearEmptyRect(bodyRect) && !nearEmptyRect(leadingRect) && !nearEmptyRect(trailingRect) && NSMaxX(trailingRect) < NSMinX(leadingRect)) {
+      highlightedPreeditPoints = [rectVertex(leadingRect) mutableCopy];
+      highlightedPreeditPoints2 = [rectVertex(trailingRect) mutableCopy];
     } else {
-      NSRectFill(stripRect);
+      highlightedPreeditPoints = [multilineRectVertex(leadingRect, bodyRect, trailingRect) mutableCopy];
+    }
+    // Expand the boxes to reach proper border
+    expand(highlightedPreeditPoints, innerBox, outerBox);
+    expand(highlightedPreeditPoints2, innerBox, outerBox);
+    highlightedPreeditPath = drawSmoothLines(highlightedPreeditPoints, 0.3*_hilitedCornerRadius, 1.4*_hilitedCornerRadius);
+    if (highlightedPreeditPoints2.count > 0) {
+      highlightedPreeditPath2 = drawSmoothLines(highlightedPreeditPoints2, 0.3*_hilitedCornerRadius, 1.4*_hilitedCornerRadius);
+    }
+  }
+
+  [NSBezierPath setDefaultLineWidth:0];
+  backgroundPath = drawSmoothLines(rectVertex(backgroundRect), _cornerRadius*0.3, _cornerRadius*1.4);
+  // Nothing should extend beyond backgroundPath
+  borderPath = [backgroundPath copy];
+  [borderPath addClip];
+  borderPath.lineWidth = _borderWidth;
+
+  // Calculate intersections
+  if (![highlightedPath isEmpty]) {
+    [backgroundPath appendBezierPath:[highlightedPath copy]];
+    if (![highlightedPath2 isEmpty]) {
+      [backgroundPath appendBezierPath:[highlightedPath2 copy]];
+    }
+  }
+  
+  if (![preeditPath isEmpty]) {
+    [backgroundPath appendBezierPath:[preeditPath copy]];
+  }
+  
+  if (![highlightedPreeditPath isEmpty]) {
+    if (preeditPath != nil) {
+      [preeditPath appendBezierPath:[highlightedPreeditPath copy]];
+    } else {
+      [backgroundPath appendBezierPath:[highlightedPreeditPath copy]];
+    }
+    if (![highlightedPreeditPath2 isEmpty]) {
+      if (preeditPath != nil) {
+        [preeditPath appendBezierPath:[highlightedPreeditPath2 copy]];
+      } else {
+        [backgroundPath appendBezierPath:[highlightedPreeditPath2 copy]];
+      }
+    }
+  }
+  [backgroundPath setWindingRule:NSEvenOddWindingRule];
+  [preeditPath setWindingRule:NSEvenOddWindingRule];
```

Comment:
When hilighted box and background color are transparent and close, sometimes there will be a noticeable gap between them, if that's undesirable, disable the lines from `// Calculate intersections` to this line.
If so, hilighted box's transparency will be blocked by background color.
This is a tradeoff to make.